qiskit/passmanager/base_optimization_tasks.py

```diff
@@ -0,0 +1,231 @@
+            state: A local state information associated with this optimization workflow.
+            callback: A callback function which is caller per execution of optimization task.
+
+        Returns:
+            Optimized passmanager IR.
+        """
+        pass
+
+
+class GenericPass(OptimizerTask, ABC):
+    """Base class of a single optimization task.
+
+    The optimization pass instance can read and write to the provided :class:`.PropertySet`.
+    """
+
+    def __init__(self):
+        self.requires: Iterable[OptimizerTask] = []
+        self._state = None
+
+    def name(self) -> str:
+        """Name of the pass."""
+        return self.__class__.__name__
+
+    def execute(
+        self,
+        passmanager_ir: Any,
+        state: PassState | None = None,
+        callback: Callable = None,
+    ) -> Any:
+        if state is not None:
+            self._state = state
+
+        for required in self.requires:
+            passmanager_ir = required.execute(
+                passmanager_ir=passmanager_ir,
+                state=self.state,
+                callback=callback,
+            )
+        if self in self.state.completed_passes:
+            self.state.previous_run = RunState.SKIP
+            return passmanager_ir
+
+        start_time = time.time()
+        ret = None
+        try:
+            ret = self.run(passmanager_ir)
+            if ret is None:
+                # Analysis pass may not return
+                ret = passmanager_ir
+        except Exception as ex:
+            self.state.previous_run = RunState.FAIL
+            raise ex
+        finally:
+            running_time = time.time() - start_time
+            log_msg = f"Pass: {self.name()} - {running_time * 1000:.5f} (ms)"
+            logger.info(log_msg)
+            if callback is not None:
+                callback(
+                    task=self,
+                    passmanager_ir=ret,
+                    property_set=self.state.property_set,
+                    running_time=running_time,
+                    count=self.state.count,
+                )
+        self.state.previous_run = RunState.SUCCESS
+        self.state.count += 1
+        self.state.completed_passes.add(self)
+        return ret
+
+    @abstractmethod
+    def run(
+        self,
+        passmanager_ir: Any,
+    ) -> Any:
+        """Run optimization task.
+
+        Args:
+            passmanager_ir: Passmanager IR to optimize.
+
+        Returns:
+            Optimized passmanager IR.
+        """
+        pass
+
+
+class BaseFlowController(OptimizerTask, ABC):
+    """Base class of flow controller.
+
+    Flow controller is built with a list of optimizer tasks, and executes them with an input
+    passmanager IR. Subclass must implement how the tasks are iterated over.
+    Note that the flow controller can be nested into another flow controller,
+    and flow controller itself doesn't provide any optimization subroutine.
+    """
+
```

Comment:
In the conventional design, running pass manager had responsibility of flow controller iteration and underlying pass execution. This introduced heavy overhead of type checking and execution logic dispatch in the running pass manager. In the proposed implementation, I redesign them with composite pattern, and I view `Task` as an interface, `GenericPass` as a leaf, and `BaseFlowController` as a composite. I prefer type safety design, and the base flow controller is a class that implements `.append` and `.pipeline`. 

> we actually already had to hack around this in the preset passmanagers when deciding which layout pass to use in the presence of control flow. pipeline and append would be API hindrances, because there's no one static pipeline.

This is very good point. However, existing `FlowControllerLinear`, `ConditionalController`, and `DoWhileController` are all iterators, so maybe just renaming the base class with `IterableFlowController` is sufficient here? Then you can write `IfElseController` which is a direct subclass of `Task`. Because outer class just calls `.execute` method, class hierarchy doesn't matter as long as your class aligns with the interface.

Reply:
I think maybe I want to take a step back and wonder about why we want to have `FlowController` at all.  Type safety (to me, at least) follows from knowing what the API invariants _must_ be, and I think that `IfElse` couldn't fit in the form proposed currently shows that we're maybe not enforcing the right things.

As you say, it would already technically be possible for `BasePass` to re-implement the flow-controller logic, since it's got everything it needs.  If so, why would we want `ConditionalController` or `DoWhileController` to be `FlowController` instances already?  We need to gain something from having separate `FlowController` logic, or we may as well just reinvent them all as direct subclasses of `Task` that have other `Task`s.  To me, what we gain is a centralised way to _run_ a series of several tasks based on custom logic for choosing the next pass to run, whereas `append` and `pipeline` are (I think?) aimed at trying to be "type safe" about _construction_ of a series of several tasks.  The problem is that by having those semantics, they constrain the logic that the "run" component can have, which is the main point of the class (as I see it).

As a concrete alternative, I'd consider defining `FlowController` to have the following properties:
- there must be a static construction method (i.e. directly from `__init__`), and its signature is dependent only on the flow controller; no restrictions from `BaseFlowController.__init__`.
- there's no standard way to get the contained passes outside the context of a run, because there's no defined order that the passes occur in.  If we need specific queries, I'd suggest methods that return / query iterators over the set of passes that _may_ run with no defined order, and that would be an abstract method.
- We might consider replacing `__iter__` with an abstract method that lets us pass in the property set / pass state explicitly?  I'm not sure about that, but for illustration I changed it just to show a way of injecting a new property set in in a way that would let `BaseFlowController` be re-entrant.
- The `execute` stays basically the same way as you have them (updated for any changes to `__iter__`).

My form of the base class would look like a slimmed down version of yours in terms of methods, and a tweaked `execute`:
```python
class BaseFlowController(Task, ABC):
    def __init__(self, options=None):
        super().__init__()
        self._options = {} if options is None else options
        self._state = None

    @abstractmethod
    def iter_tasks(self, state: PassState): Generator[Task]:
        ...

    def execute(self, ir, state, callback):
        tasks = self.iter_tasks(state)
        try:
            next_task = tasks.send(None)
        except StopIteration:
            return ir
        while True:
            ir = next_task.execute(ir, state, callback)
            state = next_task.state
            try:
                next_task = tasks.send(state)
            except StopIteration:
                break
        # Handle state propagation, timing, etc, then:
        return ir
```
Now `IfElseController` is implemented as:
```python
class IfElseController(BaseController):
    def __init__(self,
        cond_fn: Callable[[PassState], bool],
        true_tasks: Iterable[Task] = (),
        false_tasks: Iterable[Task] = (),
        *,
        options=None,
    ):
        self.cond_fn = cond_fn
        self.true_tasks = tuple(true_tasks)
        self.false_tasks = tuple(false_tasks)
        super().__init__(options)

    def iter_tasks(self, state):
        if cond_fn(state):
            yield from self.true_tasks
        else:
            yield from self.false_tasks
```
and `DoWhileController` would be:
```python
class DoWhileController(BaseController):
    def __init__(self,
        cond_fn: Callable[[PassState], bool],
        tasks: Iterable[Task] = (),
        *,
        options=None,
    ):
        self.cond_fn = cond_fn
        self.tasks = tuple(true_tasks)
        super().__init__(options)

    def iter_tasks(self, state):
        while cond_fn(state):
            for task in self.tasks:
                state = yield task
```

The `Generator` form of the iteration _might_ be unnecessary, but as in the `DoWhileController` potential implementation above, it can be used to make `FlowController`s by default re-entrant; there's no need for them to have an instance `state` attribute or even to rely on it being mutable, even if they want to query it many times because the `execute` method sends in the new state after each task run.

Reply:
Put another way: using `__iter__` gets us a "flow controller" / "running flow controller" split, except the state of the "running flow controller" still depends on the _mutable_ instance state of "flow controller".  Switching to `iter_tasks(): Generator[Task]` gets us the split, except we can pass in the new mutable state after every task, so the iteration is only dependent on the _immutable_ state of the `FlowController`, since we expect that the actual lists of passes should be fixed after the class is instantiated.

Reply:
If `append` is present more for backwards compatibility, that's something we could add only to `IfElseController`/`DoWhileController` with an immediate deprecation warning saying "these should be passed the full list of passes immediately", and `IfElseController` raises an exception if `append` is called and it's actually got two branches.

Reply:
Thanks, now your point is clear. Can you check the update in [5701447](https://github.com/Qiskit/qiskit/pull/10127/commits/5701447d6e5b01f78742a2ae077486a205007e0e)? As you say, `.append` doesn't need to be a part of the flow controller, because the role of controller is just execution of a set of tasks. Indeed your `IfElseController` class is a good example.

Regarding the example, I was also thinking the use of Generator.send was a good idea, but I remembered that `PassState` was not a local status information of a single pass. Originally these information were stored in the removed `RunningPassManager`, which was an executor of the entire task workflow. Namely, the pass state is sort of a global parameter shared among passes like a property set. This means the latest status must be set before the task is run, e.g.

```python
    def iter_tasks(self, property_set):
        state = None
        for task in self.tasks:
             if state is not None:
                 # receive previous state from generator
                 task.state = state
             state = yield task
```

for example, the `completed_passes` set from the previous task run is needed for `GenericPass.execute` to decide whether execute itself or not. So there is little gain from sending the state to generator. I also renamed `PassState` to `WorkflowStatus`.

Reply:
In my example, I was meaning that `iter_tasks` should receive as an argument anything that it's expected to be allowed to base its decisions off that might vary from pass to pass, rather than specifically what's now called `WorkflowStatus`, sorry.  In concrete terms, that means (imo) it would take the "compilation state" arguments that are given to `execute`, which are `property_set: PropertySet` and `status: WorkflowStatus` now, because it's allowed to use both of those to determine the next pass to run, but it's not allowed to use the current IR; the pipeline it's yielding from should contain a suitable analysis pass in that case.

In your example, you don't need `iter_tasks` to do `task.state = state` because that's handled by `Controller.execute` passing the state into `task.execute` as it calls it.  `iter_tasks` isn't responsible for preparing the tasks to be run - that's `execute` - it's just responsible for determining the order to return them to `execute`.  It receives the initial state in the function call, and then it receives an updated state object after each subtask runs to decide what to yield next (if it wants).

The state after the "previous" task in my example is always given to each task in a flow controller, even if it's the first task of a controller.

---

I think I might have been wrong about the splitting of the `PropertySet` and the `WorkflowStatus`, if we do do this `Generator.send` method to give us the logic split; we want the `send_type` to be a single object that allows access to its arguments by name, in case we need to update the interface in the future.  I guess the argument for splitting `WorkflowStatus` and `PropertySet` still holds, but we might need a wrapper class around both to use as the `send_type`, and I guess in that situation, it _would_ be more ergonomic to have `execute` accept that directly.